// thankshell-libs/auth.js
const AWS = require("aws-sdk")
const dynamo = new AWS.DynamoDB.DocumentClient()

exports.getUserId = async(claims) => {
    if (claims.identities) {
        const identities = JSON.parse(claims.identities)

        const authId = identities.providerName + ':' + identities.userId
        const result = await dynamo.get({
            TableName: process.env.AUTH_TABLE_NAME,
            Key:{
                'auth_id': authId,
            },
        }).promise()

        return result.Item ? result.Item['user_id'] : null
    } else {
        return claims['cognito:username']
    }
}

exports.getUser = async(userId) => {
    if (userId) {
        const result = await dynamo.get({
            TableName: process.env.USERS_TABLE_NAME,
            Key:{
                'user_id': userId,
            },
        }).promise()

        if (result.Item) {
            const {status, user_id, display_name} = result.Item
            return {
                status: status,
                user_id: user_id,
                displayName: display_name ? display_name : user_id,
            }
        }
    }

    return {
        status: 'UNREGISTERED',
        'user_id': userId,
        'displayName': userId,
    }
}

exports.getUserInfo = async(claims) => {
    const userId = await exports.getUserId(claims)

    return await exports.getUser(userId)
}
