// thankshell-libs/GroupController.js
const Auth = require('thankshell-libs/auth.js');
const appInterface = require('thankshell-libs/interface.js');
const AWS = require("aws-sdk");
const crypto = require('crypto');

const dynamo = new AWS.DynamoDB.DocumentClient();

const getGroup = async(groupId) => {
    const result = await dynamo.get({
        TableName: process.env.GROUPS_TABLE_NAME,
        Key:{
            'group_id': groupId,
        },
    }).promise()

    return result.Item
};

const getHash = (message) => {
    return crypto.createHash('sha256').update(message).digest('hex')
};

const getMemberDetails = async(memberIds, secrets, adminUser) => {
    const allMembers = await Promise.all(memberIds.map(userId => Auth.getUser(userId)))

    let dict = {}
    allMembers.forEach(user => {
        const userId = user.user_id
        const data = {
            state: user.status,
            displayName: user.displayName,
        }
        if (user.status === 'UNREGISTERED') {
            data.linkParams = {
                m: userId,
                hash: getHash(userId + secrets),
            }
        }

        dict[userId] = data
    })

    return dict
};

const Permission = {
    ADMIN: 'admin',
    MEMBER: 'member',
    VISITOR: 'visitor',
};

const getPermission = (userId, group) => {
    if (!userId || !group.members.values.includes(userId)) {
        return Permission.VISITOR
    }

    if (group.admins.values.includes(userId)) {
        return Permission.ADMIN
    }

    return Permission.MEMBER
};

module.exports = class GroupController {
  async get(authId, groupId) {
    const group = await getGroup(groupId)
    if (!group) {
        throw new appInterface.ApplicationError(`group '${groupId}' is not found`, 404)
    }

    const userId = await Auth.getMemberIdAsync(groupId, authId);
    const permission = getPermission(userId, group)

    console.log(`${userId} has ${permission} for ${groupId}`)

    const {owner, admins, bank_id, members, secrets, token_name} = group

    const publicData = {
        groupId: groupId,
        permission: permission,
        owner: owner,
    }

    if (permission === Permission.VISITOR) {
        return publicData
    }

    const memberDetails = await getMemberDetails(members.values, secrets, (permission === Permission.ADMIN))

    return {
        ...publicData,
        memberId: userId,
        admins: admins,
        bankId: bank_id,
        tokenName:  token_name ? token_name : 'coin',
        members: memberDetails,
    }
  }
};
