// thankshell-libs/GroupController.js
const GroupDao = require('thankshell-libs/GroupDao.js');
const GroupMembersDao = require('thankshell-libs/GroupMembersDao.js');
const GroupMemberPermission = require('thankshell-libs/GroupMemberPermission.js');
const appInterface = require('thankshell-libs/interface.js');
const crypto = require('crypto');


const getHash = (message) => {
  return crypto.createHash('sha256').update(message).digest('hex')
};

const getMemberDetails = async(allMembers, secrets, adminUser) => {
  let dict = {}
  allMembers.forEach(user => {
    const userId = user.user_id
    const data = {
      state: user.status,
      displayName: user.displayName,
    }
    if (user.status === 'UNREGISTERED' && adminUser) {
      data.linkParams = {
        m: userId,
        hash: getHash(userId + secrets),
      }
    }

    dict[userId] = data
  })

  return dict
};

const Permission = {
    OWNER: 'owner',
    ADMIN: 'admin',
    MEMBER: 'member',
    VISITOR: 'visitor',
};

const getPermissionString = (permission) => {
  switch (permission) {
    case GroupMemberPermission.OWNER:
      return Permission.OWNER;
    case GroupMemberPermission.ADMIN:
      return Permission.ADMIN;
    case GroupMemberPermission.MEMBER:
      return Permission.MEMBER;
    default:
      return Permission.VISITOR;
  }
};

module.exports = class GroupController {
  constructor() {
    this.groupDao = new GroupDao();
    this.groupMembersDao = new GroupMembersDao();
  }

  async get(authId, groupId) {
    console.log(`getting group with ${groupId}`);
    const group = await this.groupDao.getAsync(groupId);
    if (!group) {
      throw new appInterface.ApplicationError(`group '${groupId}' is not found`, 404)
    }

    console.log(`finding member with ${authId} from ${groupId}`);
    const member = await this.groupMembersDao.findMemberByAuthIdAsync(groupId, authId);
    if (member) {
      const permission = getPermissionString(member.permission);

      console.log(`${authId} has ${permission} for ${groupId}`)

      const allMembers = await this.groupMembersDao.getMembersAsync(groupId);

      return {
        groupId: groupId,
        permission: permission,
        owner: await this.groupMembersDao.getOwnerAsync(groupId),
        memberId: member.memberId,
        admins: await this.groupMembersDao.getAdminsAsync(groupId),
        bankId: group.bankId,
        tokenName: group.tokenName,
        members: await getMemberDetails(allMembers, group.secrets, (GroupMemberPermission.ADMIN <= member.permission)),
      }
    } else {
      console.log(`${authId} has visitor permission for ${groupId}`)

      return {
        groupId: groupId,
        permission: Permission.VISITOR,
        owner: await this.groupMembersDao.getOwnerAsync(groupId),
      }
    }
  }
};
