// thankshell-libs/interface.js
const DEFAULT_ERROR_CODE = "UNKNOWN_ERROR";

class ApplicationError extends Error {
  constructor(message, errorCode=DEFAULT_ERROR_CODE, statusCode=500) {
    super(message)
    this.statusCode = statusCode
    this.errorCode = errorCode
  }
}

class PermissionDeniedError extends ApplicationError {
  constructor(message='AccessDenied') {
    super(message, 'PERMISSION_DENIED_ERROR', 403);
  }
};

exports.ApplicationError = ApplicationError
exports.PermissionDeniedError = PermissionDeniedError;

exports.getSuccessResponse = (data={}, statusCode=200) => {
    return {
        statusCode: statusCode,
        headers: {"Access-Control-Allow-Origin": "*"},
        body: JSON.stringify(data),
    }
}

exports.getErrorResponse = (err) => {
    const { errorCode, statusCode } = (err instanceof ApplicationError) ? {
        errorCode: err.errorCode,
        statusCode: err.statusCode,
    } : {
        errorCode: DEFAULT_ERROR_CODE,
        statusCode: 500,
    }

    return {
        statusCode: statusCode,
        headers: {"Access-Control-Allow-Origin": "*"},
        body: JSON.stringify({
            code: errorCode,
            message: err.message,
        }),
    }
}
